### PR TITLE
feat: EOA bytecode and bytecode_len implementation

### PR DESCRIPTION
**Pull request by danilowhk**
Implemented EOA bytecode() and bytecode_len()

## Pull Request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [X] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Build-related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

Resolves: #210 

## What is the new behavior?

-
-
-

## Does this introduce a breaking change?

- [ ] Yes
- [X] No

### CODE REVIEW

File: crates/eoa/src/tests/test_externally_owned_account.cairo [hunk 1]
```diff
@@ -1,3 +1,57 @@
...
 
+#[cfg(test)]
+mod test_external_owned_account {
+    use eoa::externally_owned_account::{
+        IExternallyOwnedAccount, ExternallyOwnedAccount, IExternallyOwnedAccountDispatcher,
+        IExternallyOwnedAccountDispatcherTrait
+    };
+    use starknet::class_hash::Felt252TryIntoClassHash;
+    use starknet::{deploy_syscall, ContractAddress, get_contract_address, contract_address_const};
+    // Use debug print trait to be able to print result if needed.
```
Comment (enitrat) [inline crates/eoa/src/tests/test_externally_owned_account.cairo:11]: remove this comment

File: crates/eoa/src/tests/test_externally_owned_account.cairo [hunk 2]
```diff
@@ -1,3 +1,57 @@
...
+    use debug::PrintTrait;
+    use array::{ArrayTrait};
+    use traits::{Into, TryInto};
+    use result::ResultTrait;
+    use option::OptionTrait;
 
+    // Use starknet test utils to fake the transaction context.
```
Comment (enitrat) [inline crates/eoa/src/tests/test_externally_owned_account.cairo:18]: remove this comment

File: crates/eoa/src/tests/test_externally_owned_account.cairo [hunk 3]
```diff
@@ -1,3 +1,57 @@
...
+    use starknet::testing::{set_caller_address, set_contract_address};
+
+    fn deploy_external_owned_account() -> IExternallyOwnedAccountDispatcher {
```
Comment (enitrat) [inline crates/eoa/src/tests/test_externally_owned_account.cairo:21]: shorter is better 😄 

```suggestion
    fn deploy_eoa() -> IExternallyOwnedAccountDispatcher {
```

File: crates/eoa/src/tests/test_externally_owned_account.cairo [hunk 4]
```diff
@@ -1,3 +1,57 @@
...
+        IExternallyOwnedAccountDispatcher { contract_address }
+    }
+
+    #[test]
+    #[available_gas(2000000000)]
+    fn test_bytecode() {
+        let owner = contract_address_const::<1>();
+        set_contract_address(owner);
+
+        let external_owner_account_contract = deploy_external_owned_account();
```
Comment (enitrat) [inline crates/eoa/src/tests/test_externally_owned_account.cairo:38]:
```suggestion
        let eoa_contract = deploy_eoa();
```

File: crates/eoa/src/tests/test_externally_owned_account.cairo [hunk 5]
```diff
@@ -1,3 +1,57 @@
...
+
+        assert(
+            external_owner_account_contract.bytecode() == ArrayTrait::<u8>::new().span(),
```
Comment (enitrat) [inline crates/eoa/src/tests/test_externally_owned_account.cairo:41]:
```suggestion
            eoa_contract.bytecode() == ArrayTrait::<u8>::new().span(),
```

File: crates/eoa/src/tests/test_externally_owned_account.cairo [hunk 6]
```diff
@@ -1,3 +1,57 @@
...
+        );
+    }
+
+    #[test]
+    #[available_gas(2000000000)]
+    fn test_bytecode_len() {
+        let owner = contract_address_const::<1>();
+        set_contract_address(owner);
+
+        let external_owner_account_contract = deploy_external_owned_account();
```
Comment (enitrat) [inline crates/eoa/src/tests/test_externally_owned_account.cairo:52]:
```suggestion
        let eoa_contract = deploy_eoa();
```

File: crates/eoa/src/tests/test_externally_owned_account.cairo [hunk 7]
```diff
@@ -1,3 +1,57 @@
...
+        let value: u32 = 0;
+
+        assert(external_owner_account_contract.bytecode_len() == value, 'wrong bytecode');
```
Comment (enitrat) [inline crates/eoa/src/tests/test_externally_owned_account.cairo:55]:
```suggestion
        assert(eoa_contract.bytecode_len() == value, 'wrong bytecode');
```

File: crates/eoa/src/externally_owned_account.cairo [hunk 8]
```diff
@@ -1,3 +1,30 @@
...
+
+    #[storage]
+    struct Storage {
+        starknet_address: ContractAddress
+    }
+
+    #[external(v0)]
+    impl ExternallyOwnedAccount of super::IExternallyOwnedAccount<ContractState> {
+        fn bytecode(self: @ContractState) -> Span<u8> {
+            return ArrayTrait::<u8>::new().span();
```
Comment (enitrat) [inline crates/eoa/src/externally_owned_account.cairo:22]: add a TODO here reminding us to implement that

File: crates/eoa/src/externally_owned_account.cairo [hunk 9]
```diff
@@ -1,3 +1,30 @@
...
+        }
+
+        fn bytecode_len(self: @ContractState) -> u32 {
+            return 0;
+        }
```
Comment (enitrat) [inline crates/eoa/src/externally_owned_account.cairo:27]: same here

File: crates/eoa/src/externally_owned_account.cairo (hunk 8 above)
Comment (danilowhk) [reply crates/eoa/src/externally_owned_account.cairo:22]: These are needed for EXTCODE opcodes, will add comments

https://github.com/kkrt-labs/kakarot/blob/c0c80301d4dfe384247b423a3bc2726522b5d1a8/src/kakarot/accounts/eoa/externally_owned_account.cairo#L132

Comment (enitrat) [reply crates/eoa/src/externally_owned_account.cairo:23]: ah yes sorry I was confused. makes sense for an eoa

File: crates/eoa/src/externally_owned_account.cairo [hunk 10]
```diff
@@ -1,3 +1,32 @@
...
+    }
+
+    #[external(v0)]
+    impl ExternallyOwnedAccount of super::IExternallyOwnedAccount<ContractState> {
+        // @notice Empty bytecode needed for EXTCODE opcodes.
+        fn bytecode(self: @ContractState) -> Span<u8> {
+            return ArrayTrait::<u8>::new().span();
+        }
+
+        // @notice Empty bytecode needed for EXTCODE opcodes.
```
Comment (enitrat) [inline crates/eoa/src/externally_owned_account.cairo:26]: same for this

File: crates/eoa/src/tests/test_externally_owned_account.cairo [hunk 11]
```diff
@@ -1,3 +1,49 @@
...
+    }
+
+    #[test]
+    #[available_gas(2000000000)]
+    fn test_bytecode_len() {
+        let owner = contract_address_const::<1>();
+        set_contract_address(owner);
+
+        let eoa_contract = deploy_eoa();
+        let value: u32 = 0;
```
Comment (enitrat) [inline crates/eoa/src/tests/test_externally_owned_account.cairo:45]: variable here is a bit confusing, i'd remove it.

File: crates/eoa/src/tests/test_externally_owned_account.cairo [hunk 12]
```diff
@@ -1,3 +1,48 @@
...
+
+    #[test]
+    #[available_gas(2000000000)]
+    fn test_bytecode() {
+        let owner = contract_address_const::<1>();
+        set_contract_address(owner);
+
+        let eoa_contract = deploy_eoa();
+
+        assert(eoa_contract.bytecode() == ArrayTrait::<u8>::new().span(), 'wrong bytecode');
```
Comment (enitrat) [inline crates/eoa/src/tests/test_externally_owned_account.cairo:35]:
```suggestion
        assert(eoa_contract.bytecode() == Default::default().span(), 'wrong bytecode');
```

File: crates/eoa/src/tests/test_externally_owned_account.cairo [hunk 13]
```diff
@@ -1,3 +1,48 @@
...
+    }
+
+    #[test]
+    #[available_gas(2000000000)]
+    fn test_bytecode_len() {
+        let owner = contract_address_const::<1>();
+        set_contract_address(owner);
```
Comment (enitrat) [inline crates/eoa/src/tests/test_externally_owned_account.cairo:42]: mocking is useless in those tests

Comment (danilowhk) [reply crates/eoa/src/tests/test_externally_owned_account.cairo]: Removed

**Review by enitrat — APPROVED** [review]

**Pull request merged**